### PR TITLE
Use [ceil|floor]WithTolerance methods in GridExtent

### PR DESCRIPTION
**Pull request by CloudNiner**
~~Still waiting on an edge case test that allows us to consistently reproduce this issue~~, but as part of some other bugfixing/testing I updated GridExtent to use the tolerance methods consistently everywhere. This appears to have resolved the issue in the withResolution calls that caused mismatched grid extents.

They are _not_ used in `GridExtent.alignTargetPixels` because that method provides
a GDAL specific operation copied from GDAL source.

## Notes

The tolerance in GridExtent is 1e-7, in QGIS its 1e-6 and GDAL its 1e-5. Should we update our value to match one of the other projects? This could have the consequence of subtly breaking projects that rely on the current behavior for correctness. 

## Testing

Take a look at the first commit's CI build: https://app.circleci.com/pipelines/github/locationtech/geotrellis/28/workflows/e5506f7e-9024-4d1b-83b2-2a2f78945639/jobs/233

In there, you should find a test failure for the test `should compute cols and rows in GridExtents with non integer cell sizes` (use ctrl+f to find in page).

Then take a look at the HEAD of this branch's CI build: https://app.circleci.com/pipelines/github/locationtech/geotrellis/29/workflows/18bca2a1-c6b6-47d0-963c-e9ddbab9431d/jobs/246/steps

It should succeed with the new test now passing. Can search for the same passing test in the CI output.

Connects #3248

### CODE REVIEW

**Comment by pomadchin** [comment]
@CloudNiner what about math.round calls? should they still be math.rounds or it should be smth else?

**Comment by CloudNiner** [comment]
Unsure. Would probably prefer to avoid further changes until we have a test case to work against which I still haven't found.

File: raster/src/test/scala/geotrellis/raster/GridExtentSpec.scala [hunk 1]
```diff
@@ -136,5 +136,28 @@ class GridExtentSpec extends FunSpec with Matchers {
...
+        val isGood: CellSize => Boolean = (cs: CellSize) => {
+          val cellCount = width / (cs.width / baseCellSize.width) * (height / (cs.height / baseCellSize.height))
+          println(s"Cell count for $cs: $cellCount")
+          cellCount < targetCellCount
+        }
+
+        sizes.filter(isGood)
+      }
+      val path = baseGeoTiffPath("failing-withResolutions.tif")
+      val rsSb = GeoTiffRasterSource(path)
```
Comment (pomadchin) [inline raster/src/test/scala/geotrellis/raster/GridExtentSpec.scala:157]: Hm, I think you can extract `GridExtents` manually for the test case and you don't even need this test file to be used in tests at all.

File: raster/src/main/scala/geotrellis/raster/GridExtent.scala [hunk 2]
```diff
@@ -70,8 +70,8 @@ class GridExtent[@specialized(Int, Long) N: Integral](
       throw GeoAttrsError(s"illegal cellheights: $cellheight and ${that.cellheight}")
 
     val newExtent = extent.combine(that.extent)
-    val newRows = ceil(newExtent.height / cellheight)
-    val newCols = ceil(newExtent.width / cellwidth)
+    val newRows = ceilWithTolerance(newExtent.height / cellheight)
```
Comment (pomadchin) [inline raster/src/main/scala/geotrellis/raster/GridExtent.scala:73]: Should not be [here](https://github.com/locationtech/geotrellis/blob/4a021962d194b7693f8dff4b0cd9e10d49683c44/raster/src/main/scala/geotrellis/raster/GridExtent.scala#L45-L57) used ceilWithTolerance as well?

It looks like we're computing cols / rows via the ceil operation but validating it via the `round` function.

Can you also check that https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/MosaicRasterSource.scala#L160-L163 would work through the `GridExtent.combine`?

^ the last one I can check later; but if [this test](https://github.com/locationtech/geotrellis/blob/cf6e8596a0b21c4d3598525080b98dc087c8f898/gdal/src/test/scala/geotrellis/raster/gdal/GDALRasterSourceSpec.scala#L202) wont fail it is a good sign.

Comment (CloudNiner) [reply raster/src/main/scala/geotrellis/raster/GridExtent.scala]: > It looks like we're computing cols / rows via the ceil operation but validating it via the round function.

In short, yes, after a discussion with @echeipesh. And GridExtent currently operates in its constructor using `round` so in order to remain consistent, we should use that elsewhere when we construct new GridExtents, like in `withResolution` and `combine` to avoid mismatches throwing exceptions that are out of user control like in #3248. **I've updated this PR to use `round` everywhere and added an additional test to verify `withResolution`.** I've also created https://github.com/locationtech/geotrellis/issues/3261 in order to line up some future work to improve `withResolution`. I'm going to add a bit of documentation in here to describe this situation before merging as well.

> Can you also check...

I think we still need the logic there that you linked to for now because GridExtent.combine  will throw an exception unless the two GridExtents to be combined have the same CellSize, which isn't a guarantee here when constructing a mosaic.

Comment (pomadchin) [reply raster/src/main/scala/geotrellis/raster/GridExtent.scala]: sounds good to me, thanks!
